### PR TITLE
Simplify `EXPR LIKE 'constant'` to `expr = 'constant'`

### PR DESCRIPTION
**Pull request by adriangb**
https://github.com/apache/datafusion/pull/12978#issuecomment-2427600551

Closes https://github.com/apache/datafusion/issues/13192

### CODE REVIEW

**Comment by adriangb** [comment]
Wow, this is now rewriting `col ~ '^foo'` -> `col like 'foo%'` -> `starts_with(col, 'foo')` 🤯 
If we implement stats pushdown for `starts_with(col, 'foo')` we'll automatically get `col ~ '^foo'` pushed down 💪🏻

File: datafusion/optimizer/Cargo.toml
Comment (alamb) [reply datafusion/optimizer/Cargo.toml:44]: I think we need to make some interface for rewriting the function through 🤔 

Perhaps adding a method to `SimplifyInfo` like

```rust
/// return a new Expr with a function call for exact string prefix max
fn make_starts_with(base: Expr, prefix: Expr) -> Option<Expr>
```

Comment (adriangb) [reply datafusion/optimizer/Cargo.toml:44]: Hmm not sure what you mean. Is this a pattern already in use somewhere I can learn from?

Comment (alamb) [reply datafusion/optimizer/Cargo.toml:44]: I was suggest adding a method to https://docs.rs/datafusion/latest/datafusion/logical_expr/simplify/trait.SimplifyInfo.html

However i can't quite figure out how to plumb it in without getting messy

THe trick is that this rewrite is based on the semantcs of `starts_with` -- so I think logically it would need to live along side the implementation of `starts_with` because it depends on those semantics.

And similarly, to do the rewrite in the pruning predicate we need something similar (somehow to provide a per function rewrite) 🤔

Comment (adriangb) [reply datafusion/optimizer/Cargo.toml:44]: Hmm okay I'm happy to work on this more but I fear I am not familiar enough with the relationship between all of these things to figure out how to wire everything up correctly. Maybe we pause this work until someone with more familiarity of these bits of code can suggest a way to structure this?

Comment (alamb) [reply datafusion/optimizer/Cargo.toml:44]: I will try and help push this forward this week. I am sorry for the delay but I have been super busy the last few days

Comment (adriangb) [reply datafusion/optimizer/Cargo.toml:44]: No problem 😄

**Comment by alamb** [comment]
I thought more about this PR the other day -- given the dependecy you have identified, I wonder if we might actually want to do the opposite 

Specifically rewrite 
```sql
starts_with(x, 'foo')
```

into 
```sql
x LIKE 'foo%'
```

With the rationale that `like` already has quite a bit of special case / optimization. 

I'll file a ticket to track the idea

**Comment by adriangb** [comment]
Hmm interesting. `startswith` does seem simpler to me in principle, but I'd understand if it's better to do the other way around.

The dependency is a bit strange to me, I'm not sure why functions would depend on the optimizer, I would think the optimizer could depend on functions.

**Review by alamb — COMMENTED** [review]
@adriangb  what would you think about updating this PR to handle the no wild card case.

The more I think about the prefix case the less I am convinced that translating to some other function would be valuable. Instead I am thinking maybe we can ensure the `LIKE` evaluation is fast (perhaps it can detect this case and invoke a better kernel) 🤔

**Comment by adriangb** [comment]
To handle ONLY the no wildcard case? Happy to do that if that's all the progress we can make for now.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1485]: Today a downstream project can choose which functions to use, this is configurable in the session, right? The DF repertoire of functions serves as a convenient package, but not mandatory.
Here we're inserting implicit function call to a specific function.
Is this following a pre-existing pattern? if so, i am OK with this.
But maybe it would be better to eg consult session state whether this function was supposed to be used. What do you think?

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1449]: keep `expr` as `expr` 
use `like_expr` when referring to the whole `Expr::Like` thing

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1459]: This is cool idea, but `is_null(like_expr)` checks makes it incorrect.
This function returns "is this always null". Applied to a column reference, it will return false, so this optimization will kick in, but the column values are still _nullable_.

The "more correct" way would be to

- optimize `WHERE col LIKE '%'` to `WHERE col IS NOT NULL`
- optimize `WHERE col NOT LIKE '%'` to `WHERE false`.

However, this simplification cannot be done **here**. It can only be done for top-level filter conjuncts, or when expr being simplified is immediately used as a condition (eg `IF(expr, ...)`). In other cases, this need to be NULL-preserving boolean logic:

```
trino> SELECT a, a LIKE '%', a NOT LIKE '%'
    -> FROM (VALUES NULL, '', 'x') t(a);
  a   | _col1 | _col2
------+-------+-------
 NULL | NULL  | NULL
      | true  | false
 x    | true  | false
(3 rows)
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1463]: Honor `escape_char`

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1482]: `&& underscore_index.is_none()`

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:3638]: add test cases

- with `%` pattern 
- with `_` pattern (can be simplified to a length check, but that can be follow up)

- with `%%` and `%%%` patterns (can be simplified just like single `%`)
- with `__` and `___` patterns (can be simplified to a length check, but that can be follow up)
- with patterns combining % and _, eg `a_b%`, `a%b_` (these cannot simplified) and  `_%`,  `%_`  (these can be simplified to a length check, but that can be follow up)

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1459]: The easiest solution would be to just drop this special case.
I.e. `expr LIKE '%'` would get simplified to `starts_with(expr, '')`

**Review by findepi — CHANGES_REQUESTED** [review]

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1459]: See also https://github.com/apache/datafusion/issues/12637#issuecomment-2449352017

**Comment by findepi** [comment]
do we have equivalent of https://github.com/apache/datafusion/pull/12978 for starts_with?

**Comment by adriangb** [comment]
> do we have equivalent of #12978 for starts_with?

No but I think it will have a lot of code sharing with #12978 and should be relatively easy to implement. After we merge that PR I'd be happy to go back in and implement it, doing whatever refactoring is needed to get good code sharing.

**Comment by adriangb** [comment]
@alamb @findepi I redid this PR now to only handle the equality cases and close the issue.

I do still think it'd be nice to figure out how to rewrite to startswith but it seems like that requires some larger scaler refactoring that I don't have the knowledge or time to carry out.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (adriangb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1478]: This code was already here (see diff above). Did I change the logic in any way?

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (adriangb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1480]: I made it not simplify if there is an escape character for now

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (adriangb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:3656]: Added in 6c847fa6a

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1478]: Ack. Can you please confirm or disconfirm the correctness concern?

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (findepi) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1481]: nit: can this condition be checked on the same line with `!pattern_str.contains(['%', '_']`
it would reduce code branching.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (adriangb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1478]: I agree this is incorrect but this is the logic that was there before this PR. I'm happy to remove it. I guess no one has complained because the case that breaks is `SELECT NULL LIKE '%'` which isn't something that happens under normal usage.

Comment (findepi) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs]: > I agree this is incorrect but this is the logic that was there before this PR.

thanks for confirming and understood it's pre-existing.

Can it be fixed though?
i personally would prefer that `git annotate` on the file wouldn't point at me as the author of something that's known to be incorrect.

Comment (adriangb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs]: My latest commit removes that special case. I'm not sure how to fix it, your comment suggests it can't be fixed right?

Comment (findepi) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs]: > I guess no one has complained because the case that breaks is `SELECT NULL LIKE '%'` which isn't something that happens under normal usage.

actually the opposite.
this was correct only for this case
and wrong for all other `... LIKE '%'` queries.

see https://github.com/apache/datafusion/pull/13259

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs
Comment (adriangb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1485]: awesome, I'll wait for your work to be merged then continue here

**Comment by adriangb** [comment]
Closing in favor of #13260

**Pull request closed without merging**
